spring-kafka/src/main/java/org/springframework/kafka/listener/adapter/MessagingMessageListenerAdapter.java

```diff
@@ -293,28 +307,130 @@ private String evaluateTopic(Object request, Object source, Object result, Expre
 		}
 		else {
 			Object value = sendTo.getValue(this.evaluationContext, new ReplyExpressionRoot(request, source, result));
-			Assert.state(value instanceof String, "replyTopic expression must evaluate to a String or Address");
+			boolean isByteArray = value instanceof byte[];
+			if (value == null) {
+				value = tryTopicFallback(source);
+				if (value != null) {
+					isByteArray = true;
+				}
+			}
+			if (!(value instanceof String || isByteArray)) {
+				throw new IllegalStateException(
+					"replyTopic expression must evaluate to a String or byte[], it is: "
+					+ (value == null ? null : value.getClass().getName()));
+			}
+			if (isByteArray) {
+				return new String((byte[]) value, StandardCharsets.UTF_8);
+			}
 			return (String) value;
 		}
 	}
 
-	@SuppressWarnings("unchecked")
+	/**
+	 * Send the result to the topic.
+	 *
+	 * @deprecated in favor of {@link #sendResponse(Object, String, Object)}.
+	 * @param result the result.
+	 * @param topic the topic.
```

Comment:
Shoudn't `@deprecated` tag be in the end of JavaDocs?

```
<module name="AtclauseOrder">
		<property name="target" value="METHOD_DEF, CTOR_DEF, VARIABLE_DEF"/>
		<property name="tagOrder" value="@param, @return, @throws, @since, @deprecated, @see"/>
</module>
```

Hm... That is our Checkstyle config here. How has your change bypassed it?